### ISSUE TITLE
[Bug]: Missing required parameter in EKS role_arn

### ISSUE THREAD

**Issue by fabiovelenopuro** [hashicorp/terraform-provider-aws#27036]
### Terraform Core Version

1.3.1

### AWS Provider Version

4.5.0

### Affected Resource(s)

after updating the terraform version from **1.3.0** to **1.3.1** I get this error when running the plan

```
│ Error: Missing required argument
│ 
│   with aws_eks_cluster.orca-eks-cluster-dev,
│   on EKS-cluster1-dev.tf line 9, in resource "aws_eks_cluster" "eks-cluster-dev":
│    9:     role_arn = aws_iam_role.iam-eks-cluster_role.arn
│ 
│ The argument "role_arn" is required, but no definition was found.
```

the EKS-cluster1-dev.tf is this:

```
resource "aws_eks_cluster" "eks-cluster-dev" {
    name     = "eks-cluster-dev"
    role_arn = aws_iam_role.iam-eks-cluster_role.arn
   ....
```

and the resource iam-eks-cluster_role is this:

```
resource "aws_iam_role" "iam-eks-cluster_role" {
  name = "iam-eks-cluster_role"
  assume_role_policy = jsonencode({
    Statement = [{
      Action = "sts:AssumeRole"
      Effect = "Allow"
      Principal = {
        AWS = [
            "arn:aws:iam::XXXXXXXXXX:user/devops-user"
        ],
        Service = [
            "eks.amazonaws.com"
            ]
      }
    }]
    Version = "2012-10-17"
  })
}

```

### Expected Behavior

I expect the role_arn parameter works correctly and recognise the associated role:
with this syntax:
`role_arn = aws_iam_role.iam-eks-cluster_role.arn
` 

### Actual Behavior

after updating the terraform version from **1.3.0** to **1.3.1** I get this error when running the plan

```
│ Error: Missing required argument
│ 
│   with aws_eks_cluster.orca-eks-cluster-dev,
│   on EKS-cluster1-dev.tf line 9, in resource "aws_eks_cluster" "eks-cluster-dev":
│    9:     role_arn = aws_iam_role.iam-eks-cluster_role.arn
│ 
│ The argument "role_arn" is required, but no definition was found.
```

### Relevant Error/Panic Output Snippet

```shell
│ Error: Missing required argument
│ 
│   with aws_eks_cluster.orca-eks-cluster-dev,
│   on EKS-cluster1-dev.tf line 9, in resource "aws_eks_cluster" "eks-cluster-dev":
│    9:     role_arn = aws_iam_role.iam-eks-cluster_role.arn
│ 
│ The argument "role_arn" is required, but no definition was found.
```

### Terraform Configuration Files

```
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
      version = ">= 3.59.0"
    }
  }
  required_version = ">= 1.0.7"
}
```

### Steps to Reproduce

create a EKS resource like this:

```
resource "aws_eks_cluster" "eks-cluster-dev" {
    name     = "eks-cluster-dev"
    role_arn = aws_iam_role.iam-eks-cluster_role.arn
    vpc_config {
      subnet_ids = [aws_subnet.sbn-eks-cl-pub-az_a.id, aws_subnet.sbn-eks-cl-pub-az_b.id, aws_subnet.sbn-eks-cl-pub-az_c.id]
    }
    enabled_cluster_log_types = ["api", "audit", "authenticator", "controllerManager", "scheduler"]
}

resource "aws_iam_role" "iam-eks-cluster_role" {
  name = "iam-eks-cluster_role"
  lifecycle {
      ignore_changes = all
    }
  assume_role_policy = jsonencode({
    Statement = [{
      Action = "sts:AssumeRole"
      Effect = "Allow"
      Principal = {
        AWS = [
            "arn:aws:iam::XXXXXXXXXXXX:user/devops-user"
        ],
        Service = [
            "eks.amazonaws.com"
            ]
      }
    }]
    Version = "2012-10-17"
  })
}
```

### Debug Output

```
│ Error: Missing required argument
│ 
│   with aws_eks_cluster.orca-eks-cluster-dev,
│   on EKS-cluster1-dev.tf line 9, in resource "aws_eks_cluster" "eks-cluster-dev":
│    9:     role_arn = aws_iam_role.iam-eks-cluster_role.arn
│ 
│ The argument "role_arn" is required, but no definition was found.
```

**Comment by mattburgess** [comment]
I'm unable to reproduce this using Terraform 1.3.1 and AWS Provider 4.33.0:

Here's my entire TF code:

```hcl
data "aws_availability_zones" "current" {}

resource "aws_vpc" "eks" {
  cidr_block           = "192.168.0.0/16"
  enable_dns_hostnames = true
  enable_dns_support   = true
}

resource "aws_subnet" "private" {
  count             = length(data.aws_availability_zones.current.names)
  vpc_id            = aws_vpc.eks.id
  availability_zone = data.aws_availability_zones.current.names[count.index]
  cidr_block        = cidrsubnet(aws_vpc.eks.cidr_block, 8, count.index)
}

resource "aws_eks_cluster" "example" {
  name     = "example"
  role_arn = aws_iam_role.eks_cluster.arn

  vpc_config {
    subnet_ids = aws_subnet.private.*.id
  }

  depends_on = [
    aws_iam_role_policy_attachment.eks-cluster-AmazonEKSClusterPolicy,
  ]
}

data "aws_iam_policy_document" "eks_cluster_assume_role" {
  statement {
    effect = "Allow"
    principals {
      type        = "Service"
      identifiers = ["eks.amazonaws.com"]
    }
    actions = ["sts:AssumeRole"]

  }
}

resource "aws_iam_role" "eks_cluster" {
  name               = "eks-cluster"
  assume_role_policy = data.aws_iam_policy_document.eks_cluster_assume_role.json
}

resource "aws_iam_role_policy_attachment" "eks-cluster-AmazonEKSClusterPolicy" {
  policy_arn = "arn:aws:iam::aws:policy/AmazonEKSClusterPolicy"
  role       = aws_iam_role.eks_cluster.name
}
```

A plan shows:

```sh
Plan: 7 to add, 0 to change, 0 to destroy.
```

The only way I can get the same error as you report is by commenting out or deleting that `role_arn` line.

Can you confirm that you're still seeing the same issue and whether or not you see the same problem with the above TF code?

**Comment by mattburgess** [comment]
Also, I just spotted that your error message states:

```sh
│ Error: Missing required argument
│ 
│   with aws_eks_cluster.orca-eks-cluster-dev,

But your provided code snippet doesn't configure that cluster resource, it's configuring `eks-cluster-dev`. If you can still reproduce this, then showing the TF code for the `orca-eks-cluster-dev` cluster resource might help us work out what's going on.

**Comment by fabiovelenopuro** [comment]
hi @mattburgess 

as for the "orca" prefix, I had cleaned up the code of prefixes to anonymise it, but I forgot to clean up the output, in any case it refers to the resource: **eks-cluster-dev**, sorry for mistake.

**TEST 1 with terraform version 1.3.0**

telling you more about the life cycle of the code, it was written about 4-5 months ago with earlier versions of terraform, and up to 1.3.0 (to which I downgraded the last update) the output is this:

No changes. Your infrastructure matches the configuration.

```
Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
```
here the terrafrom version:

```
$ terraform version
Terraform v1.3.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v4.5.0

Your version of Terraform is out of date! The latest version
is 1.3.1. You can update by downloading from https://www.terraform.io/downloads.html
```

**TEST 2 with Terraform version 1.3.1**
just now, I reinstalled the new version of terraform from the hashicorp repository:
```
Terraform v1.3.1
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v4.5.0
```

and running the terraform plan I get the same error:

 ```
Error: Missing required argument
│ 
│   with aws_eks_cluster.eks-cluster-dev,
│   on EKS-cluster1-dev.tf line 8, in resource "aws_eks_cluster" "eks-cluster-dev":
│    8:     role_arn = aws_iam_role.iam-eks-cluster_role.arn
│ 
│ The argument "role_arn" is required, but no definition was found.
```

There seems to be a problem with the **role_arn** parameter which has this value in my code:

`role_arn = aws_iam_role.iam-eks-cluster_role.arn
`

**TEST 3 with Terraform version 1.3.1 and the arn as a string in the code**

using Terraform 1.3.1 I tried also putting the arn of the AWS role as the value of role_arn directly 
`role_arn = "arn:aws:iam::XXXXXXXXXXXX:role/iam-eks-cluster_role"`
and the plan gives me this output:

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:


  # aws_iam_role.iam-eks-cluster_role will be updated in-place
  ~ resource "aws_iam_role" "iam-eks-cluster_role" {
      - arn                   = "arn:aws:iam::XXXXXXXXXXXX:role/iam-eks-cluster_role" -> null
      - create_date           = "2022-03-15T18:00:05Z" -> null
        id                    = "iam-eks-cluster_role"
        name                  = "iam-eks-cluster_role"
        tags                  = {}
      - unique_id             = "YYYYYYYYYYYYYYYYYYYYY" -> null
        # (6 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 2 to change, 0 to destroy.
```

**TEST 4 however, if I downgrade to terraform 1.3.0** 

and re-run the plan with:
`role_arn: "arn:aws:iam::XXXXXXXXXXXX:role/iam-eks-cluster_role" `
or with:
`role_arn: aws_iam_role.iam-eks-cluster_role.arn`
i get, as expected:
`No changes. Your infrastructure matches the configuration.`

**My conclusion**
which is why I suppose there is a problem in the new version of Terraform 1.3.1 for this type of parameter: **role_arn**


I hope I have been clear in my explanation.

thank you very much for your support

**Comment by RobertKeyser** [comment]
For what it's worth, I'm having the same issue with `aws_lb_listener_certificate` and `listener_arn` on `v1.3.1` of Terraform and `4.2.0` of the AWS provider.

```
│ Error: Missing required argument
│ 
│   with aws_lb_listener_certificate.cert,
│   on alb.tf line 64, in resource "aws_lb_listener_certificate" "cert":
│   64:   listener_arn    = aws_lb_listener.listener.arn
│ 
│ The argument "listener_arn" is required, but no definition was found.
```

Similar setup, where the `listener_arn` is set in the resource, but it's not being recognized. Downgrading the Terraform version makes it work again.

**Comment by mattburgess** [comment]
I appreciate the fact that the Terraform version upgrade/downgrade would strongly hint that the bug is on that side of things rather than with the AWS provider. However, if either of you are in a position to upgrade the provider to the latest version so we can rule that out please, I'd appreciate it.

In addition, Terraform v1.3.2  was released today. None of the fixes look remotely related to this issue though, and I also couldn't see any related issues in Terraform's github repo. But, if you're able to reproduce with the latest Terraform and latest AWS provider that would definitely be helpful.

**Comment by fabiovelenopuro** [comment]
> I appreciate the fact that the Terraform version upgrade/downgrade would strongly hint that the bug is on that side of things rather than with the AWS provider. However, if either of you are in a position to upgrade the provider to the latest version so we can rule that out please, I'd appreciate it.
> 
> In addition, Terraform v1.3.2 was released today. None of the fixes look remotely related to this issue though, and I also couldn't see any related issues in Terraform's github repo. But, if you're able to reproduce with the latest Terraform and latest AWS provider that would definitely be helpful.

Hi @mattburgess ,
thanks for the news, I immediately ran the update to terraform 1.3.2 and tested it by doing a plan:

```
No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and found no differences, so no changes are needed.
$ terraform version
Terraform v1.3.2
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v4.5.0
```

**evrything works fine!**
the issue seems resolved in terraform version 1.3.2

Thanks
F

**Comment by mattburgess** [comment]
@fabiovelenopuro - thanks for confirming!
@justinretzolk - would you mind closing this one please? Thanks!
